File: 144 preorderTraversal/144(1).c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* preorderTraversal(struct TreeNode* root, int* returnSize){
    int *ans=(int*)malloc(2000*sizeof(int));
    *returnSize=0;
    if(root==NULL){
        return ans;
    }

    struct TreeNode* stk[2000];
    struct TreeNode* node=root;
    int stk_top=0;
    while(stk_top>0 || node!=NULL){
        while(node!=NULL){
            ans[(*returnSize)++]=node->val;
            stk[stk_top++]=node;
            node=node->left;
        }
        node=stk[--stk_top];
        node=node->right;
    }
    return ans;
}

File: 144 preorderTraversal/144(1).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans=[]
        if not root:
            return ans
        stack=[]
        node=root
        while stack or node:
            while node:
                ans.append(node.val)
                stack.append(node)
                node=node.left
            node=stack.pop()
            node=node.right
        return ans    